### PR TITLE
73-prepare-dataset-for-the-llama-fine-tuning-and-perform-an-example-fine-tune-with-it

### PR DESCRIPTION
**Pull request by GustawB**
This PR introduces code for fine-tuning the llama models as well as libraries used for generating datasets for the fine-tuning. Those libraries are ready to be used in other places as well, there is an example usage for them and finally, they are covered by tests.

### CODE REVIEW

**Review by SzymonKozl — CHANGES_REQUESTED** [review]
GJ in overall.
please include requirements for this code in reasonable places.
As we are starting to submit tests alongside code, here are 2 suggestions from me. This is not obligatory to implement is this PR and lets talk about this on tomorrows call:
* test edge cases, not only basic flow
* for opening files mocking can be used - see [this](https://stackoverflow.com/questions/1289894/how-do-i-mock-an-open-used-in-a-with-statement-using-the-mock-framework-in-pyth)

File: llama_finetuning/fine_tune_llama.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            warmup_steps=10,
+            output_dir=dataset_name,
+            seed=0,
+            report_to="wandb"
+        ),
+    )
+
+    trainer.train()
+
+@app.function(image=finetune_image, gpu="H100", timeout=600)
```
[the remark this answers is not in the document]
Comment (GustawB) [reply llama_finetuning/fine_tune_llama.py:101]: Done

File: src/llama_dataset_generation/datasetter.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    """
+    This function takes a pandas dataframe and a mapping function, and returns a Dataset object
+    with the mapping function applied to it (it adds a prompt to the data).
+
+    :param df: A pandas dataframe to be mapped.
+    :param map_func: A mapping function to be applied to the dataframe.
+    :return training_data: A Dataset object with the mapping function applied.
+    """
+    if map_func in [one_input_multiple_outputs_wrequest, one_input_multiple_outputs_wthrequest]:
+        training_data = Dataset.from_pandas(__parse_to_oimo(df))
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/llama_dataset_generation/datasetter.py]: Done

[the remark this answers is not in the document]
Comment (GustawB) [reply src/llama_dataset_generation/datasetter.py]: Reverted

File: src/llama_dataset_generation/ground_truth_parser.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    This function initializes the Async OpenAI client.
+
+    :param api_key: The OpenAI API key.
+    :return client: The initialized Async OpenAI client
+    """
+    client = AsyncOpenAI(api_key=api_key)
+    return client
+
+
+def store_coupons_as_json(coupons: json, path: string) -> None:
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/llama_dataset_generation/ground_truth_parser.py:21]: Nope, it should be a list. Fixed.

File: src/llama_dataset_generation/ground_truth_parser.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    :param coupons_list: The list of coupon data.
+    :param client: The initialized Async OpenAI client.
+    :param batch_size: The size of each batch.
+    :return res: The extracted discount details.
+    """
+    tasks = []
+    for j in range(0, len(coupons_list), batch_size):
+        task = asyncio.create_task(
+            __get_data(__get_prompt(coupons_list[j:min(j + 15, len(coupons_list))]), client, model='gpt-4o'))
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/llama_dataset_generation/ground_truth_parser.py:121]: fixed

File: src/llama_dataset_generation/ground_truth_parser.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                Return only json data, without any annotations or additional text.
+                Do not skip any entries."""
+    return prompt
+
+
+async def __extract_discount_details(coupons_list: list, client: AsyncOpenAI, batch_size: int = 15) -> list:
+    """
+    This function handles the logic of creating async tasks to process the coupon data in batches.
+
+    :param coupons_list: The list of coupon data.
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/llama_dataset_generation/ground_truth_parser.py:113]: Fair point. Done

**Comment by GustawB** [comment]
Added requirements for each logical component

File: src/llama_dataset_generation/input_parser.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import string
+
+import pandas as pd
+
+__X_COLUMN = 'text'
+__Y_COLUMN = 'seen_timestamp'
+
+
+def __concat_column_x_by_column_y(column_x: string, column_y: string, df: pd.DataFrame) -> dict:
```
[the remark this answers is not in the document]
Comment (GustawB) [reply src/llama_dataset_generation/input_parser.py]: Done
